src/cheri_pte.sail

```diff
@@ -78,23 +78,57 @@ bitfield PTE_Bits : pteAttribs = {
 /*
  * Reserved PTE bits used by extensions on RV64.
  *
- * There are no such bits on RV32/SV32, so we hard-code both CapWrite=1
- * and CapRead=1 (i.e., there is no ability to disable capability loads
- * from or stores to pages in the SV32 MMU).
+ * For capability store the behaviors are as follows.
+ *
+ *   CW  CD  Action
+ *
+ *    0   X    Trap on tagged capability store
+ *    1   0    CAS the PTE to CW=1, CD=1 (i.e., set both bits) or trap
+ *    1   1    Permit tagged capability store
+ *
+ * The CW/CD pair is akin to Sv{32,48,...}'s W/D bits.
+ *
+ * The CW=1 CD=0 behavior is described as a CAS (and not an AMO OR) of the PTE
+ * to close race conditions, much as with W and D.  Should the TLB have this
+ * state cached and then observe a CW=1 CD=1 PTE, no PTE write is necessary.
+ * On the other hand, if CW=0 is observed, the store operation must trap.
+ *
+ * The intended semantics for capability loads are as follows.
+ *
+ *   CR  CRT CRG Action
+ *
+ *    0   0   0  Capability loads strip tags
+ *    0   1   0  Capability loads trap (on set tag)
+ *    0   X   1  [Reserved]
+ *
+ *    1   0   0  Capability loads succeed: no traps or tag clearing
+ *    1   0   1  [Reserved]
+ *    1   1   G  Capability loads trap if G mismatches [ms]ccsr.[su]gclg, where
+ *                   the compared bit is keyed off of this PTE's U.  sccsr is
+ *                   used if the machine has Supervisor mode, otherwise mccsr.
+ *
+ * Sv32: There are no extension bits available, so we hard-code the result to
+ * CW=1 CR=1 CD=1 CRT=0 CRG=0
  */
 type extPte = bits(10)
 
 bitfield Ext_PTE_Bits : extPte = {
   CapWrite     : 9, /* Permit capability stores */
```

Comment:
This is pre-existing, but bits 9 and 8 conflict with “Svpbmt” Standard Extension for Page-Based Memory Types. I wonder if we should shift everything down by two as part of this change series or start from zero (since upstream started from the top)?

Reply:
Ouch.  I guess that's not the worst possible flag day, it'd just touch sail, the cores, and the CheriBSD kernel.